### ISSUE TITLE
Cache lock problem

### ISSUE THREAD

**Issue by jeanpralo** [Mashape/kong#2186]
Hi there,

I am developing a custom plugin and try to use the cache, but I am having an issue with it:

```
2017/03/10 01:23:17 [error] 62#0: *125 [lua] database_cache.lua:153: get_or_set(): failed to acquire cache lock: timeout, client: 10.66.100.1, server: kong, request: "GET /toto HTTP/1.1"
```

Here is a quick snippet of the code I have:

```
local function load_role(role)
   local dao_role, err = singletons.dao.sh_roles:find_all {role = role}

   if #dao_role ~= 1 or err then
      return responses.send(401, "Bad token, not a valid role")
   end

   return dao_role[1]
end

local function do_check_roles(roles)
   local verb = ngx.req.get_method()
   local uri = ngx.var.uri

   for k,role in pairs(roles) do
      local dao_role = cache.get_or_set("roles."..role, nil, load_role, role)
      --dao_role = load_role(role)                                                                                                                                                                                                         

      for ruri,verbs in pairs(dao_role.access) do
         if string.match(uri, '^' .. ruri .. '$') then
            if has_value(verbs, verb) then
               return true
            end
         end
      end
   end

   return false, {status = 401, message = "Permission denied"}
end

```


The error does not happen straight away but only seems to be triggered after the third time and only in the case where the callback is returning directly a 401 response.

### Steps To Reproduce

1. Add an api
2. Map a custom plugin to this api
3. Query kong about two or three times and that should trigger the bug

### Additional Details & Logs

- Kong 0.10.0 (docker image)


Logs:

```
2017/03/10 01:27:17 [debug] 61#0: *68 [lua] base_plugin.lua:17: access():  executing plugin "roles-filtering-plugin": access
2017/03/10 01:27:17 [debug] 61#0: *68 [lua] base_plugin.lua:21: header_filter():  executing plugin "roles-filtering-plugin": header_filter
2017/03/10 01:27:17 [debug] 61#0: *68 [lua] base_plugin.lua:25: body_filter():  executing plugin "roles-filtering-plugin": body_filter
2017/03/10 01:27:17 [debug] 61#0: *68 [lua] base_plugin.lua:25: body_filter():  executing plugin "roles-filtering-plugin": body_filter
2017/03/10 01:27:17 [debug] 61#0: *68 [lua] base_plugin.lua:29: log():  executing plugin "roles-filtering-plugin": log
2017/03/10 01:27:17 [info] 61#0: *68 client 10.66.100.1 closed keepalive connection
2017/03/10 01:27:19 [debug] 61#0: *129 [lua] base_plugin.lua:17: access():  executing plugin "roles-filtering-plugin": access
2017/03/10 01:27:19 [info] 61#0: *133 client 127.0.0.1 closed keepalive connection
2017/03/10 01:27:19 [info] 61#0: *131 client 127.0.0.1 closed keepalive connection
2017/03/10 01:27:19 [info] 61#0: *134 client 127.0.0.1 closed keepalive connection
2017/03/10 01:27:20 [debug] 62#0: *152 [lua] cluster.lua:20: log(): [cluster] auto-joining
2017/03/10 01:27:22 [info] 61#0: *161 client 127.0.0.1 closed keepalive connection
2017/03/10 01:27:22 [info] 61#0: *162 client 127.0.0.1 closed keepalive connection
2017/03/10 01:27:22 [info] 61#0: *165 client 127.0.0.1 closed keepalive connection
2017/03/10 01:27:23 [debug] 62#0: *182 [lua] cluster.lua:20: log(): [cluster] auto-joining
2017/03/10 01:27:24 [error] 61#0: *129 [lua] database_cache.lua:153: get_or_set(): failed to acquire cache lock: timeout, client: 10.66.100.1, server: kong, request: "GET /toto HTTP/1.1", host: "10.66.100.101"
2017/03/10 01:27:24 [error] 61#0: *129 lua entry thread aborted: runtime error: /plugins/kong/plugins/roles-filtering-plugin/handler.lua:93: attempt to index local 'dao_role' (a nil value)
stack traceback:
coroutine 0:
	/plugins/kong/plugins/roles-filtering-plugin/handler.lua: in function 'do_check_roles'
	/plugins/kong/plugins/roles-filtering-plugin/handler.lua:123: in function 'access'
	/usr/local/share/lua/5.1/kong.lua:256: in function 'access'
	access_by_lua(nginx-kong.conf:81):2: in function <access_by_lua(nginx-kong.conf:81):1>, client: 10.66.100.1, server: kong, request: "GET /toto HTTP/1.1", host: "10.66.100.101"
2017/03/10 01:27:25 [info] 61#0: *192 client 127.0.0.1 closed keepalive connection
2017/03/10 01:27:25 [info] 61#0: *191 client 127.0.0.1 closed keepalive connection




stack traceback:
coroutine 0:
	/plugins/kong/plugins/roles-filtering-plugin/handler.lua: in function 'do_check_roles'
	/plugins/kong/plugins/roles-filtering-plugin/handler.lua:123: in function 'access'
	/usr/local/share/lua/5.1/kong.lua:256: in function 'access'
	access_by_lua(nginx-kong.conf:81):2: in function <access_by_lua(nginx-kong.conf:81):1>, client: 10.66.100.1, server: kong, request: "GET /toto HTTP/1.1", host: "10.66.100.101"

```

**Comment by Tieske** [comment]
the callback gets locked/unlocked, so if you exit early, it will never unlock the cache lock. So the error checking must be outside of the callback.

**Comment by jeanpralo** [comment]
sweet got it ! Thanks !

**Comment by jeanpralo** [comment]
@Tieske  I guess that would be worth noting that some default plugin do that at the  moment:

https://github.com/Mashape/kong/blob/master/kong/plugins/jwt/handler.lua#L117

https://github.com/Mashape/kong/blob/master/kong/plugins/jwt/handler.lua#L61

Also I have removed this return from the cb and I still have this cache lock problem. Note that it only happens when I have no value to cache because there is no result:

```
local function load_role(role)
   local dao_role, err = singletons.dao.sh_roles:find_all {role = role}

   if #dao_role ~= 1 or err or not dao_role[1] then
      return false, {status = 401, message = "Bad token, not a valid role"}
   end

   return true, dao_role[1]
end

local function do_check_roles(roles)
   local verb = ngx.req.get_method()
   local uri = string.gsub(ngx.var.request_uri, "?.*", "")

   for k,role in pairs(roles) do
      local err, dao_role = cache.get_or_set("roles."..role, nil, load_role, role)

      if not err then
         responses.send(dao_role.status, dao_role.message)
      end

      for ruri,verbs in pairs(dao_role.access) do
         print('uri:'..uri..', match: '..ruri)
         if string.match(uri, '^' .. ruri .. '$') then
            if has_value(verbs, verb) then
               return true
            end
         end
      end
   end

   return false, {status = 401, message = "Permission denied"}
end

```

**Comment by Tieske** [comment]
yes, this I realised as well. Checking the existing codebase was already on my todo list. See https://github.com/Mashape/kong/commit/77e58d1fd1f52a4cc3a7af742a6de64a54aba61e

**Comment by jeanpralo** [comment]
Fair enough :) 

Would you have any idea why with that code I pasted above I still have those deadlock then ? Since I return every time from the cb.

**Comment by jeanpralo** [comment]
@Tieske  I think I know why, this is because of that piece of code here:

https://github.com/Mashape/kong/blob/77e58d1fd1f52a4cc3a7af742a6de64a54aba61e/kong/tools/database_cache.lua#L167

So in the end the cb must always return one param except in case of a major error ! So it is either the queried value or nil. Would that be a fair assumption ?

**Comment by Tieske** [comment]
thx for bringing this to my attention. The problem seems bigger than anticipated. The code you point at has the same problem. It exits early, as the lock is only released further down.

Working on a fix right now.

**Comment by Tieske** [comment]
see #2197

**Comment by Tieske** [comment]
related #2166

**Comment by rocketspacer** [comment]
Can you merge the fix to kong **0.9.x**
